File: hemin-engine/src/main/java/io/hemin/engine/exception/HeminException.java
package io.hemin.engine.exception;

public class HeminException extends Exception {

    public HeminException() { }

    public HeminException(String message) {
        super(message);
    }

    public HeminException(String message, String errorConstant) {
        super(message);
    }

    public HeminException(String message, Throwable cause) {
        super(message, cause);
    }

    public HeminException(String message, String errorConstant, Throwable cause) {
        super(message, cause);
    }

    public HeminException(Throwable cause) {
        super(cause);
    }

    public HeminException(String message, Throwable cause, boolean enableSuppression, boolean writableStackTrace) {
        super(message, cause, enableSuppression, writableStackTrace);
    }

}

File: hemin-engine/src/main/scala/hemin/engine/HeminException.scala
package hemin.engine

object HeminException {
  /** Allows to access the message and the cause via pattern-matching */
  def unapply(e: HeminException): Option[(String,Throwable)] = Some((e.getMessage, e.getCause))
}

/**
  *
  * @see The class design follows the recommendations from this
  *      post [[https://stackoverflow.com/a/43942163/9812594]]
  */
class HeminException(message: String)
  extends Exception(message) {

  def this(message: String, cause: Throwable) = {
    this(message)
    initCause(cause)
  }

  def this(cause: Throwable) = this(Option(cause).map(_.toString).orNull, cause)

  def this() = this(null: String)

}
